distribution/sdk/sample-nsili-client/src/main/java/org/codice/alliance/nsili/client/SampleNsiliClient.java

```diff
@@ -270,65 +292,84 @@ public int getHitCount(Query query) throws Exception {
+    public void printDagAttributes(DAG dag) {
         LOGGER.info("--------------------");
         LOGGER.info("PRINTING DAG ATTRIBUTES");
-        for (int i = 0; i < dag.nodes.length; i++) {
-            Node node = dag.nodes[i];
+        for (Node node : dag.nodes) {
             if (node.node_type.equals(NodeType.ATTRIBUTE_NODE)) {
-                String name = node.attribute_name;
-                String value = CorbaUtils.getNodeValue(node.value);
-                LOGGER.info("{} = {}", name, value);
+                LOGGER.info("{} = {}", node.attribute_name, CorbaUtils.getNodeValue(node.value));
             }
         }
         LOGGER.info("--------------------");
     }
 
-    public void retrieveProductFromDAG(DAG dag) throws MalformedURLException {
+    public void downloadProductFromDAG(DAG dag) throws MalformedURLException {
         LOGGER.info("Downloading products...");
-        for (int i = 0; i < dag.nodes.length; i++) {
-            Node node = dag.nodes[i];
-            if (node.attribute_name.equals("productUrl")) {
-
-                String url = node.value.extract_string();
-                URL fileDownload = new URL(url);
-                String productPath = "product.jpg";
-                LOGGER.info("Downloading product : {}", url);
+
+        for (Node node : dag.nodes) {
+            if (node.attribute_name.equals(NsiliConstants.PRODUCT_URL)) {
+                URL fileDownloadUrl = getEncodedUrlFromString(node.value.extract_string());
+
+                final String productPath = "product.jpg";
+                LOGGER.info("Downloading product : {}", fileDownloadUrl);
                 try (FileOutputStream outputStream = new FileOutputStream(new File(productPath));
-                        BufferedInputStream inputStream = new BufferedInputStream(fileDownload.openStream())) {
+                        BufferedInputStream inputStream = new BufferedInputStream(fileDownloadUrl.openStream())) {
 
                     byte[] data = new byte[1024];
                     int count;
                     while ((count = inputStream.read(data, 0, 1024)) != -1) {
                         outputStream.write(data, 0, count);
                     }
 
-                    LOGGER.info("Successfully downloaded product from {}.", url);
+                    LOGGER.info("Successfully downloaded product from {}.", fileDownloadUrl);
                     Files.deleteIfExists(Paths.get(productPath));
-
                 } catch (IOException e) {
-                    LOGGER.error("Unable to download product from {}.", url, e);
+                    LOGGER.error("Unable to download product from {}.", fileDownloadUrl, e);
                 }
             }
         }
     }
 
-    public PackageElement[] order(ORB orb, DAG[] dags) throws Exception {
+    private URL getEncodedUrlFromString(String urlString) throws MalformedURLException {
+        URL url = new URL(urlString);
+        try {
+            URI uri = new URI(url.getProtocol(),
```

Comment:
Could we validate that all the URL pieces are set instead of using the `URI` constructor? Or if that doesn't work, could we simply call `url.toURI()`? It just looks strange that we need to create a `URI` from the `URL` pieces and then convert that `URI` back to a `URL`.